### DISCUSSION TITLE
Transform imputed data wide to long and convert into mids object by as.mids()

### DISCUSSION THREAD

**Discussion by ADoggett** [amices/mice#607] [Q&A]
Hi Stef & Team, 

I’m facing an issue with as.mids() that I am hoping you can help with. 

I’m running mice on a repeated measures dataset in the wide format, because I want to use the same variables measured at different time points as auxiliary variables. I am following the example workflow of this from [Section 11.2 of FIMD](https://stefvanbuuren.name/fimd/sec-fdd.html). My issue arises when I try to covert back to a mids object for plotting and analysis, I get a duplicate row names error. I know people have posted about getting a rownames error with as.mids before, but I have not found a post that details a solution. 

I know where the issue is arising; in lines 111-114 of the source code for as.mids, row names are reassigned:
https://github.com/cran/mice/blob/94c89a24f0456d90cf37a15be758c6304818a67d/R/as.R#L111-L114

```R
if (!is.na(.id)) {
    rownames(ini$data) <- unlist(long[imps == 0, .id], use.names = FALSE)
  }
```
If my dataset was not longitudinal, or still in wide format, there is no problem because all the .id variables are unique within imps==0. However with repeated measures data in the long format, this line causes the function to fail out because there are going to be repeated ids within imps==0. 

A crude solution of course is removing this chunk of the source code which redefines the row names, and this does appear to work for me. However I’m aware thats a bit of a dicey solution, and I feel like I am probably missing something here. What is the purpose of redefining rownames inside the as.mids() function? Is my solution of circumventing this valid or am I off track and should be following a different workflow?

Reprex provided below. 

Thanks so much in advance, 

Amanda 

```R
library(mice)
#> 
#> Attaching package: 'mice'
#> The following object is masked from 'package:stats':
#> 
#>     filter
#> The following objects are masked from 'package:base':
#> 
#>     cbind, rbind
datawide <- data.frame("id" = (c(1:100)),
                       "x" = sample(1:2,100, replace = TRUE), #some baseline variable
                       "y.1" = round(c(rnorm(100,10,2))), #outcome at time 1
                       "y.2" = round(c(rnorm(100,10,2)))) #outcome at time 2

dat_miss <- c(sample(1:100, 30, replace=FALSE))
for(i in dat_miss){
  datawide$y.2[i] <- NA
} 

impIni <- mice(datawide,print=FALSE, maxit=0)
updateMatrix <- impIni$predictorMatrix
updateMatrix["id",] <- 0
updateMatrix[,"id"] <- 0
updateMatrix["x", ] <- 0
updateMatrix["y.1", ] <- 0

imp.wide <- mice(datawide,m=5,method=impIni$method, predictorMatrix = updateMatrix, print=FALSE)
longWide <- mice::complete(imp.wide,"long",include=TRUE) #imputed datasets are stacked, but y vars are not
longLong <- reshape(longWide, idvar = ".id",
                    varying = list(c("y.1", 
                                     "y.2")),
                    direction = "long",
                    v.names = c("y"),
                    new.row.names = 1:(nrow(longWide)*2),
                    sep=".") #now truly long format with imputations and y vars stacked 
longLong <- longLong[order(longLong$.imp, longLong$.id, longLong$time),]  
row.names(longLong) <- 1:nrow(longLong) 

midsLong<- as.mids(longLong) # results in duplicate rownames error
#> Warning: non-unique values when setting 'row.names': '1', '2', '3', '4', '5',
#> '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19',
#> '20', '21', '22', '23', '24', '25', '26', '27', '28', '29', '30', '31', '32',
#> '33', '34', '35', '36', '37', '38', '39', '40', '41', '42', '43', '44', '45',
#> '46', '47', '48', '49', '50', '51', '52', '53', '54', '55', '56', '57', '58',
#> '59', '60', '61', '62', '63', '64', '65', '66', '67', '68', '69', '70', '71',
#> '72', '73', '74', '75', '76', '77', '78', '79', '80', '81', '82', '83', '84',
#> '85', '86', '87', '88', '89', '90', '91', '92', '93', '94', '95', '96', '97',
#> '98', '99', '100'
#> Error in `.rowNamesDF<-`(x, value = value): duplicate 'row.names' are not allowed
```

**Comment by stefvanbuuren** [comment]
You are changing the structure of the data created by `mice::complete()` and expect that `as.mids()` can convert those transformed data into a proper `mids` object. Function `as.mids()` was not designed to handle this situation. A lot may go wrong, and the error message is evidence of this.

For your case, try to evade the error by replacing `row.names(longLong) <- 1:nrow(longLong) ` by
```
longLong$.id <- rep(1:200, 6)
```

Also try a simpler version

```
longLong$.id <- NULL
```

Does that give you the result you desire?
